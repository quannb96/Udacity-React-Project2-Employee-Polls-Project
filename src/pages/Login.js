import { useDispatch, useSelector } from "react-redux";
import { selectAuthState, handleLogin } from "../reducers/authedSlice";
import { useState } from "react";
import { TextField, Typography, Box, Button, Grid } from "@mui/material";

const Login = () => {
  const auth = useSelector(selectAuthState);
  const dispatch = useDispatch();
  const [username, setUsername] = useState("sarahedo");
  const [password, setPassword] = useState("password123");

  const handleUsername = (e) => {
    const value = e.target.value;
    setUsername(value);
  };

  const handlePassword = (e) => {
    const value = e.target.value;
    setPassword(value);
  };

  const handleSubmit = (e) => {
    e.preventDefault();

    if (typeof username !== "string" || typeof password !== "string") {
      return;
    }
    dispatch(handleLogin({ username, password }));
    setUsername("");
    setPassword("");
  };

  return (
    <div className="App">
      <Typography variant="h2" sx={{ m: 2 }}>
        Login
      </Typography>
      <form onSubmit={handleSubmit}>
        <Grid
          container
          spacing={0}
          direction="column"
          alignItems="center"
          justifyContent="center"
        >
          <Box sx={{ m: 2 }} width={400}>
            <TextField
              label="User"
              variant="outlined"
              value={username}
              onChange={handleUsername}
              id="username"
              name="username"
              placeholder="User"
              autoComplete="username"
              autoFocus
              required
              fullWidth
              inputProps={{ "data-testid": "user-input" }}
            />
          </Box>
          <Box sx={{ m: 2 }} width={400}>
            <TextField
              label="Password"
              value={password}
              onChange={handlePassword}
              required
              name="password"
              type="password"
              id="password"
              placeholder="Password"
              autoComplete="current-password"
              fullWidth
              inputProps={{ "data-testid": "password-input" }}
            />
          </Box>
        </Grid>

        {auth.status === "failed" && (
          <p>The authentication was not successful.</p>
        )}
        <Button type="submit" variant="contained" data-testid="login-button">
          Login
        </Button>
      </form>
    </div>
  );
};

export default Login;
