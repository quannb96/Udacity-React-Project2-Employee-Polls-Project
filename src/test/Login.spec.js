import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import Login from "../pages/Login";
import authReducer from "../reducers/authedSlice";
import pollReducer from "../reducers/pollSlice";
import userReducer from "../reducers/userSlice";

const store = configureStore({
  reducer: {
    auth: authReducer,
    poll: pollReducer,
    user: userReducer,
  },
});

test("Login component matches snapshot", () => {
  const { asFragment } = render(
    <Provider store={store}>
      <Login />
    </Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});

test("Login component changes state after typing in username field", () => {
  render(
    <Provider store={store}>
      <Login />
    </Provider>
  );

  const usernameInput = screen.getByTestId("user-input");
  const passwordInput = screen.getByTestId("password-input");
  const loginButton = screen.getByTestId("login-button");

  fireEvent.change(usernameInput, { target: { value: "newUsername" } });

  expect(usernameInput.value).toBe("newUsername");

  fireEvent.change(passwordInput, { target: { value: "newPassword" } });

  expect(passwordInput.value).toBe("newPassword");

  fireEvent.click(loginButton);
});
