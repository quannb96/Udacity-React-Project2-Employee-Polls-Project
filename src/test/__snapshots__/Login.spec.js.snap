// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <h2
      class="MuiTypography-root MuiTypography-h2 css-1qkcgnr-MuiTypography-root"
    >
      Login
    </h2>
    <form>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-l6hioc-MuiGrid-root"
      >
        <div
          class="MuiBox-root css-12edpzt"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-focused Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for="username"
              id="username-label"
            >
              User
              <span
                aria-hidden="true"
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                autocomplete="username"
                class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                data-testid="user-input"
                id="username"
                name="username"
                placeholder="User"
                required=""
                type="text"
                value="sarahedo"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-14lo706"
                >
                  <span>
                    User *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-12edpzt"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for="password"
              id="password-label"
            >
              Password
              <span
                aria-hidden="true"
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                autocomplete="current-password"
                class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                data-testid="password-input"
                id="password"
                name="password"
                placeholder="Password"
                required=""
                type="password"
                value="password123"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-14lo706"
                >
                  <span>
                    Password *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sghohy-MuiButtonBase-root-MuiButton-root"
        data-testid="login-button"
        tabindex="0"
        type="submit"
      >
        Login
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </form>
  </div>
</DocumentFragment>
`;
